stumpy/core.py

```diff
@@ -3956,7 +3956,6 @@ def _mdl(disc_subseqs, disc_neighbors, S, n_bit=8):
 
 @njit(
     # "(i8, i8, f8[:, :], f8[:], i8, f8[:, :], i8[:, :], f8)",
-    parallel=True,
```

Comment:
So, where's the a parallelization happening? I looked at the code but do not see `prange` anywhere and it should be somewhere. Otherwise, `mstump` would be slow

Reply:
I guess `compute_multi_PI` cannot be parallel because the for-loop relies on the previous iteration (i.e., each iteration of the for loop is NOT independent of the previous step). Also, the part that needs to be "fast" and parallelized is in `compute_multi_D`

Reply:
In `compute_multi_PI`:

```
D_prime[:] = 0.0
    for i in range(d):
        D_prime = D_prime + np.power(D[i], 1.0 / p)

        min_index = np.argmin(D_prime)
        pos = idx - range_start
        I[i, pos] = min_index
        P[i, pos] = D_prime[min_index] / (i + 1)
        if np.isinf(P[i, pos]):  # pragma nocover
            I[i, pos] = -1
```

technically, each column of `D_prime` can be updated in parallel like:

```
D_prime[:] = 0.0
    for i in range(d):
        for j in prange(len(D_prime):
            D_prime[j] = D_prime[j] + np.power(D[i, j], 1.0 / p)

        min_index = np.argmin(D_prime)
        pos = idx - range_start
        I[i, pos] = min_index
        P[i, pos] = D_prime[min_index] / (i + 1)
        if np.isinf(P[i, pos]):  # pragma nocover
            I[i, pos] = -1
```

although, I'm not sure if it's worth it or if it's even "faster" than the vectorized version

Reply:
@NimaSarajpoor Would you mind testing this (`prange` version) and comparing it against the speed of `parallel=False`? Also, do the tests pass for `mstump` and `maamp` for this `prange` version?

Reply:
> So, where's the a parallelization happening? I looked at the code but do not see `prange` anywhere and it should be somewhere

see the inner for-loop: https://github.com/TDAmeritrade/stumpy/blob/main/stumpy/mstump.py#L896-L899



> I guess `compute_multi_PI` cannot be parallel because the for-loop relies on the previous iteration (i.e., each iteration of the for loop is NOT independent of the previous step)

Correct!

> technically, each column of D_prime can be updated in parallel like:
...
although, I'm not sure if it's worth it or if it's even "faster" than the vectorized version

Correct! Maybe we  compare the performance [for different length of `T`]

> @NimaSarajpoor Would you mind testing this (prange version) and comparing it against the speed of parallel=False? 

So, just to confirm: 
You meant `prange` version [over columns] in `core._compute_multi_PI`. Correct? And, then I should compare it with the vectorized approach [with `parallel==False`].

Reply:
> You meant prange version [over columns] in core._compute_multi_PI. Correct? And, then I should compare it with the vectorized operation [with parallel==False, which is basically the existing approach].

Yes, exactly! 

> Maybe we compare the performance [for different length of T]

Maybe test different length T and a range of dimensions (i.e., number of time series) `range(1, 11)`